hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanP.java

```diff
@@ -0,0 +1,229 @@
+                    ctx.getSerializationService()
+            );
+        }
+
+        @Override
+        public Object[] next() {
+            while (iterator.tryAdvance()) {
+                if (--remainingBatch == 0) {
+                    // limit the number of rows scanned in one call to not break the cooperative contract
+                    return null;
+                }
+                Object[] row = prepareRow(
+                        iterator.getKey(),
+                        iterator.getKeyData(),
+                        iterator.getValue(),
+                        iterator.getValueData()
+                );
+                if (row != null) {
+                    return row;
+                }
+            }
+            done = true;
+            return null;
+        }
+
+        /**
+         * Prepare the row for the given key and value:
+         * 1) Check filter
+         * 2) Extract projections
+         *
+         * @param rawKey       key as object, might be null
+         * @param rawKeyData   key as data, might be null
+         * @param rawValue     value as object, might be null
+         * @param rawValueData value as data, might be null
+         * @return Row that is ready for processing by parent operators or {@code null} if the row hasn't passed the filter.
+         */
+        private Object[] prepareRow(Object rawKey, Data rawKeyData, Object rawValue, Data rawValueData) {
+            row.setKeyValue(rawKey, rawKeyData, rawValue, rawValueData);
+
+            // Filter.
+            if (filter != null && TernaryLogic.isNotTrue(filter.evalTop(row, ctx))) {
+                return null;
+            }
+
+            Object[] row = new Object[projections.size()];
+
+            for (int j = 0; j < projections.size(); j++) {
+                row[j] = evaluate(projections.get(j), this.row, ctx);
+            }
+
+            return row;
+        }
+    }
+
+    public static ProcessorMetaSupplier onHeapMapScanP(@Nonnull MapScanMetadata mapScanMetadata) {
+        return new OnHeapMapScanMetaSupplier(mapScanMetadata);
+    }
+
+    private static final class OnHeapMapScanMetaSupplier implements ProcessorMetaSupplier {
+        private final MapScanMetadata mapScanMetadata;
+        private transient PartitionService partitionService;
+
+        private OnHeapMapScanMetaSupplier(@Nonnull MapScanMetadata mapScanMetadata) {
+            this.mapScanMetadata = mapScanMetadata;
+        }
+
+        @Override
+        public void init(@Nonnull Context context) throws Exception {
+            this.partitionService = context.jetInstance().getHazelcastInstance().getPartitionService();
+        }
+
+        @Nonnull
+        @Override
+        public Function<? super Address, ? extends ProcessorSupplier> get(@Nonnull List<Address> addresses) {
+            Set<Partition> partitions = partitionService.getPartitions();
+            Map<Address, List<Integer>> partitionsByMember = Util.assignPartitions(
```

Comment:
This is not needed as as soon as there's no member handling some partition, the `MapScanExecIterator` will fail. But I'm leaving as a preparation for remote partition reading.